cpp/ql/src/semmle/code/cpp/Declaration.qll

```diff
@@ -203,6 +203,13 @@ abstract class Declaration extends Locatable, @declaration {
    */
   Type getTemplateArgument(int index) { none() }
 
+  /**
+   * Gets the `i`th template argument value used to instantiate this declaration
+   * from a template. When called on a template, this will return the `i`th template
+   * parameter value if it exists.
```

Comment:
I don't understand what this is supposed to return when called on an uninstantiated template. For example, what does `getTemplateArgumentValue` return for the uninstantiated form of a function like this?
```cpp
template<int i>
int addToSelf() { return i + i; }
```

The `i` parameter in the uninstantiated template is perhaps similar to a `constexpr` variable, but `getTemplateArgumentValue` returns a `Expr` rather than a `Variable`.

Reply:
See the tests I have just pushed in this commit: 23604a5

Reply:
So for the `template<int i> int addToSelf() { return i + i; }` getTemplateArgumentValue returns "Unknown Literal" (I know how to get that into a known literal).

For an instantiation of addToSelf (for instance addToSelf<10>). Then `getTemplateArgumentValue` will return the value (10 in the example).

This (will be) similar behaviour to type templates.  For the prototype instantiation the getTemplateArgument returns `T` but when instantiatied it returns the type T has been given (`int` say).

In particular `i` is not (in my mind at least) a Variable, but rather just a placeholder for an expression much as T is a type placeholder in `template<typename T>`.

Reply:
The test results suggest that a lot of clarification is needed on the QLDoc for both `getTemplateArgumentValue` and `getATemplateArgument`; the latter now seems to give the _type_ of a non-type template argument.